File: DP/1575.py
class Solution:
    MOD = 10 ** 9 + 7

    def countRoutes(self, locations: List[int], start: int, finish: int, fuel: int) -> int:
        n = len(locations)
        MOD = 10**9+7
        dp = [[None] * (fuel + 1) for _ in range(n)]
        
        return self.helper(locations, start, finish, fuel, dp)

    def helper(self, locations, city, end, remainFuel, dp):
        if remainFuel < 0:
            return 0
        if dp[city][remainFuel] is not None:
            return dp[city][remainFuel]
        ans = 1 if city == end else 0

        for next_city in range(len(locations)):
            if next_city != city and remainFuel >= abs(locations[next_city] - locations[city]):
                ans = (ans + self.helper(locations, next_city, end, remainFuel - abs(locations[next_city] - locations[city]), dp)) % self.MOD
        dp[city][remainFuel] = ans
        
        return ans

File: DP/1575.cpp
class Solution {
public:
    const int MOD=1000000007;

    int countRoutes(vector<int>& locations, int start, int finish, int fuel) {
        int n = locations.size();
        vector<vector<int>> dp(n, vector<int>(fuel+1, -1));

        return helper(locations, start, finish, fuel, dp); 
    }
    int helper(vector<int>& locations, int curr, int finish, int fuel, vector<vector<int>>& dp) {
        if (fuel < 0) return 0;
        if (dp[curr][fuel] != -1) {
            return dp[curr][fuel];
        }
        int ans = curr == finish ? 1 : 0;

        for (int next = 0; next < locations.size(); next++) {
            if (next != curr && fuel >= abs(locations[next] - locations[curr])) {
                ans = (ans + helper(locations, next, finish, fuel - abs(locations[next] - locations[curr]), dp)) % MOD ;
            }
        }        
        dp[curr][fuel] = ans;
        
        return ans;
    }
};
